src/amount.rs

```diff
@@ -166,21 +204,48 @@ impl Amount {
             (-21, "zepto"),
             (-24, "yocto"),
         ]
-        .iter()
-        .cloned()
-        .collect();
+        .into_iter()
+        .collect()
+    }
+
+    /// generates an SI string for this amount, eg: "253 yotta".
+    //
+    // SI units obtained from:
+    //  http://www.knowledgedoor.com/2/units_and_constants_handbook/power_prefixes.html
+    //
+    // note: There is no SI name for 10^0, so those are just formatted
+    //       as regular integer values via ::to_string()
+    // note: SI names only extend from 10^-24..10^24.  Outside that range
+    //       we fallback to formatting with ::to_string()
+    pub fn to_si_string(self) -> String {
+        let map = Self::si_map();
+        // note: this unwrap_or can never fail because map always contains values.
+        // note: we use max+2 to account for eg 10 yotta, 100 yotta.
+        let min = map.keys().min().unwrap_or(&0);
+        let max = map.keys().max().unwrap_or(&0) + 2;
+
+        //   25 giga  = 25 * 10^9,   count = 25, unit =  9
+        //  250 giga  = 25 * 10^10,  count = 25, unit = 10
+        // 2500 giga  = 25 * 10^11,  count = 25, unit = 11
 
-        if self.unit >= -24 && self.unit <= 24 && self.count != 0 {
+        if self.unit >= *min && self.unit <= max && self.count != 0 {
             let mut unit = self.unit;
             loop {
                 if let Some(name) = map.get(&unit) {
-                    let diff = self.unit.abs() - unit.abs();
-                    let udiff = 10u64.pow(diff as u32);
-                    let newcount = self.count as u64 * udiff;
+                    let diff = (self.unit.abs() - unit.abs()).abs();
```

Comment:
I think you want this instead:
```suggestion
                    let diff = (self.unit - unit).abs();
```

Reply:
done.